styx-scheduler-service/src/test/java/com/spotify/styx/docker/KubernetesDockerRunnerTest.java

```diff
@@ -570,18 +587,58 @@ public void shouldNotRunIfSecretHasManagedServiceAccountKeySecretNamePrefix() th
     verify(pods, never()).create(any(Pod.class));
   }
 
-  @Test
-  public void shouldCompleteWithStatusCodeOnSucceeded() throws Exception {
-    createdPod.setStatus(terminated("Succeeded", 20, null));
+  @Parameters({
+      "Running,   20, true",
+      "Running,   1,  true",
+      "Running,   0,  true",
+      "Succeeded, 20, true",
+      "Succeeded, 1,  true",
+      "Succeeded, 0,  true",
+      "Failed,    20, true",
+      "Failed,    1,  true",
+      "Failed,    0,  true",
+      "Running,   20, false",
+      "Running,   1,  false",
+      "Running,   0,  false",
+      "Succeeded, 20, false",
+      "Succeeded, 1,  false",
+      "Succeeded, 0,  false",
+      "Failed,    20, false",
+      "Failed,    1,  false",
+      "Failed,    0,  false",
+  })
+  @Test
+  public void shouldCompleteWithStatusCodeOnMainContainerTerminated(String phase, int code,
+      boolean withKeepaliveContainer) throws Exception {
+    final String executionId = createdPod.getMetadata().getName();
+
+    final PodStatus podStatus = podStatusNoContainer(phase);
+
+    podStatus.getContainerStatuses()
+        .add(new ContainerStatusBuilder()
+            .withState(terminatedContainerState(code, ""))
+            .withName(executionId)
+            .build());
+
+    // Verify that old pods without a keepalive container are also correctly handled
+    if (withKeepaliveContainer) {
+      podStatus.getContainerStatuses()
+          .add(new ContainerStatusBuilder()
+              .withName(KEEPALIVE_CONTAINER_NAME)
+              .withNewState().withNewRunning().endRunning().endState()
+              .build());
+    }
+
+    createdPod.setStatus(podStatus);
     podWatcher.eventReceived(Watcher.Action.MODIFIED, createdPod);
 
     verify(stateManager).receive(Event.started(WORKFLOW_INSTANCE), -1);
-    verify(stateManager).receive(Event.terminate(WORKFLOW_INSTANCE, Optional.of(20)), 0);
+    verify(stateManager).receive(Event.terminate(WORKFLOW_INSTANCE, Optional.of(code)), 0);
   }
 
   @Test
   public void shouldFailOnErrImagePull() throws Exception {
-    createdPod.setStatus(waiting("Pending", "ErrImagePull"));
+    KubernetesPodEventTranslatorTest.setWaiting(createdPod, "Pending", "ErrImagePull");
```

Comment:
static import `setWaiting`, the same for the following.

Reply:
👍